RecoTracker/PixelSeeding/plugins/alpaka/CAHitNtuplet.cc

```diff
@@ -20,57 +22,230 @@
+          startingPairs_(iConfig.getParameter<std::vector<int>>("startingPairs")),
+          phiCuts_(iConfig.getParameter<std::vector<int>>("phiCuts")),
+          minZ_(iConfig.getParameter<std::vector<double>>("minZ")),
+          maxZ_(iConfig.getParameter<std::vector<double>>("maxZ")),
+          maxR_(iConfig.getParameter<std::vector<double>>("maxR")) {}
+
+    // Layers params
+    const std::vector<double> caThetaCuts_;
+    const std::vector<double> caDCACuts_;
+
+    // Cells params
+    // TODO: move to unsigned int here
+    const std::vector<int> pairGraph_;
+    const std::vector<int> startingPairs_;
+    const std::vector<int> phiCuts_;
+    const std::vector<double> minZ_;
+    const std::vector<double> maxZ_;
+    const std::vector<double> maxR_;
+
+    mutable edm::ESGetToken<TrackerGeometry, TrackerDigiGeometryRecord> tokenGeometry_;
+    mutable edm::ESGetToken<TrackerTopology, TrackerTopologyRcd> tokenTopology_;
+  };
+
+}  // namespace reco
+
 namespace ALPAKA_ACCELERATOR_NAMESPACE {
+
   template <typename TrackerTraits>
-  class CAHitNtupletAlpaka : public stream::EDProducer<> {
-    using HitsConstView = TrackingRecHitSoAConstView<TrackerTraits>;
-    using HitsOnDevice = TrackingRecHitsSoACollection<TrackerTraits>;
-    using HitsOnHost = TrackingRecHitHost<TrackerTraits>;
+  class CAHitNtupletAlpaka
+      : public stream::EDProducer<edm::GlobalCache<::reco::CAGeometryParams>,
+                                  edm::RunCache<cms::alpakatools::MoveToDeviceCache<Device, ::reco::CAGeometryHost>>> {
+    using HitsConstView = ::reco::TrackingRecHitConstView;
+    using HitsOnDevice = reco::TrackingRecHitsSoACollection;
+    using HitsOnHost = ::reco::TrackingRecHitHost;
 
-    using TkSoAHost = TracksHost<TrackerTraits>;
-    using TkSoADevice = TracksSoACollection<TrackerTraits>;
+    using TkSoAHost = ::reco::TracksHost;
+    using TkSoADevice = reco::TracksSoACollection;
 
     using Algo = CAHitNtupletGenerator<TrackerTraits>;
 
+    using CAGeometryCache = cms::alpakatools::MoveToDeviceCache<Device, ::reco::CAGeometryHost>;
+    using Rotation = SOARotation<float>;
+    using Frame = SOAFrame<float>;
+
   public:
-    explicit CAHitNtupletAlpaka(const edm::ParameterSet& iConfig);
+    explicit CAHitNtupletAlpaka(const edm::ParameterSet& iConfig, const ::reco::CAGeometryParams* iCache);
     ~CAHitNtupletAlpaka() override = default;
+
     void produce(device::Event& iEvent, const device::EventSetup& es) override;
+
+    static void globalEndJob(::reco::CAGeometryParams const*) { /* Do nothing */ };
+    static void globalEndRun(edm::Run const& iRun,
+                             edm::EventSetup const&,
+                             RunContext const* iContext) { /* Do nothing */ };
+
     static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
 
+    static std::shared_ptr<CAGeometryCache> globalBeginRun(edm::Run const& iRun,
```

Comment:
This is called for every run, but from a quick glance it depends only on the Event Setup.
Wouldn't it make more sense to move this to a dedicated `ESProducer` ?

Reply:
AFAICT the `iCache` members are initialized from the configuration in `initializeGlobalCache()` (except the `ESGetToken`s in the constructor). The earlier discussion on the approach is in https://github.com/cms-sw/cmssw/pull/47611#discussion_r2030160710.

Reply:
On this, to follow up on the discussion there. The idea was to simplify the building of the `CAHitNtuplet` producer without having a specific ESProducer for a new set of parameters. An intermediate way to do this could be to have:
- the CAGeometry params, so those really depending on the tracker geometry/topology event setups, filled in an ESProducer;
- the CA params (so phicuts, zmax etc.) filled here with the `MoveToDevice` + `RunCache` schema;
Most probably I would leave this, eventually, to another PR. I'd prefer this PR converge and to build on top of it if all the other pieces are in place. Also, I'd like to understand if this "third way" really simplify things or not.